apollo-runtime/src/main/java/com/apollographql/apollo/subscription/WebSocketSubscriptionTransport.kt

```diff
@@ -0,0 +1,166 @@
+      return
+    }
+    socket.send(message.serializeToJson())
+  }
+
+  internal fun onOpen() {
+    callback.onConnected()
+  }
+
+  internal fun onMessage(message: OperationServerMessage?) {
+    callback.onMessage(message)
+  }
+
+  internal fun onFailure(t: Throwable?) {
+    try {
+      callback.onFailure(t)
+    } finally {
+      release()
+    }
+  }
+
+  internal fun onClosed() {
+    try {
+      callback.onClosed()
+    } finally {
+      release()
+    }
+  }
+
+  private fun release() {
+    webSocketListener.getAndSet(null)?.release()
+    webSocket.set(null)
+  }
+
+  private fun OperationClientMessage.serializeToJson(): String =
+      toJsonString(
+        writePayloadAsJsonString = writePayloadAsJsonString,
+        extensions = extensions
+    )
+
+  internal class WebSocketListener(delegate: WebSocketSubscriptionTransport) : okhttp3.WebSocketListener() {
+    private val delegateRef = WeakReference(delegate)
+
+    override fun onOpen(webSocket: WebSocket, response: Response) {
+      delegateRef.get()?.onOpen()
+    }
+
+    override fun onMessage(webSocket: WebSocket, text: String) {
+      delegateRef.get()?.onMessage(OperationServerMessage.fromJsonString(text))
+    }
+
+    override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
+      delegateRef.get()?.onFailure(t)
+    }
+
+    override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
+      delegateRef.get()?.onClosed()
+    }
+
+    override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
+      delegateRef.get()?.onClosed()
+    }
+
+    fun release() {
+      delegateRef.clear()
+    }
+  }
+
+  /**
+   * @param webSocketUrl The URL of the GraphQL WebSocket API
+   * @param webSocketConnectionFactory The [WebSocket.Factory] to use to create the websockets
+   * @param writePayloadAsJsonString If the payload should be serialized as a JSON String (default false). Set this to `true` for AWS
+   *                                 AppSync integrations.
+   * @param extensions Any additional extensions to when sending start commands. Must not be changed after this creating the factory.
+   */
+  class Factory @JvmOverloads constructor(
+      webSocketUrl: String,
+      private val webSocketConnectionFactory: WebSocket.Factory,
+      private val writePayloadAsJsonString: Boolean = false,
+      private val extensions: Map<String, Any?> = emptyMap()
+  ) : SubscriptionTransport.Factory {
+    private val webSocketRequest: Request = Request.Builder()
+        .url(webSocketUrl)
+        .addHeader("Sec-WebSocket-Protocol", "graphql-ws")
+        .addHeader("Cookie", "")
+        .build()
+
+    override fun create(callback: SubscriptionTransport.Callback): SubscriptionTransport =
+        WebSocketSubscriptionTransport(webSocketRequest, webSocketConnectionFactory, callback, writePayloadAsJsonString, extensions)
+  }
+
+  class AppSyncFactory @JvmOverloads constructor(
```

Comment:
This was created as a convenience to do two things for the user:
* Add the `header` and `payload` query parameters which AppSync requires
* Add the authorization as an extension to the start messages.